application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/pages/internal/MethodWithFrameTypeLabelProvider.java

```diff
@@ -0,0 +1,146 @@
+	private final Image protectedJitMethodImage;
+	private final Image privateJitMethodImage;
+
+	public MethodWithFrameTypeLabelProvider() {
+		nonoptimizedMethodImage = CoreImages.METHOD_NON_OPTIMIZED.createImage();
+
+		defaultMethodImage = CoreImages.METHOD_DEFAULT.createImage();
+		publicMethodImage = CoreImages.METHOD_PUBLIC.createImage();
+		protectedMethodImage = CoreImages.METHOD_PROTECTED.createImage();
+		privateMethodImage = CoreImages.METHOD_PRIVATE.createImage();
+
+		// Construct the versions of the images for the "non-interpreted" methods (normally JITed)
+		defaultJitMethodImage = new OverlayImageDescriptor(CoreImages.METHOD_DEFAULT, false,
+				CoreImages.METHOD_JITOVERLAY).createImage();
+		publicJitMethodImage = new OverlayImageDescriptor(CoreImages.METHOD_PUBLIC, false, CoreImages.METHOD_JITOVERLAY)
+				.createImage();
+		protectedJitMethodImage = new OverlayImageDescriptor(CoreImages.METHOD_PROTECTED, false,
+				CoreImages.METHOD_JITOVERLAY).createImage();
+		privateJitMethodImage = new OverlayImageDescriptor(CoreImages.METHOD_PRIVATE, false,
+				CoreImages.METHOD_JITOVERLAY).createImage();
+	}
+
+	@Override
+	public Font getFont(Object key) {
+		return JFaceResources.getFontRegistry().getBold(JFaceResources.DEFAULT_FONT);
+	}
+
+	@Override
+	public Image getImage(Object key) {
+		if (key instanceof MethodWithFrameType) {
+			MethodWithFrameType mwft = (MethodWithFrameType) key;
+			IMCMethod method = mwft.getMethod();
+			if (key instanceof IMCMethod) {
+				method = (IMCMethod) key;
+			} else if (key instanceof IMCFrame) {
+				method = ((IMCFrame) key).getMethod();
+			}
+			if (method != null) {
+				if ((method.getModifier() & Modifier.PUBLIC) != 0) {
+					return mwft.getFrameType() != Type.INTERPRETED ? publicJitMethodImage : publicMethodImage;
+				} else if ((method.getModifier() & Modifier.PROTECTED) != 0) {
+					return mwft.getFrameType() != Type.INTERPRETED ? protectedJitMethodImage : protectedMethodImage;
+				} else if ((method.getModifier() & Modifier.PRIVATE) != 0) {
+					return mwft.getFrameType() != Type.INTERPRETED ? privateJitMethodImage : privateMethodImage;
+				}
+				return mwft.getFrameType() != Type.INTERPRETED ? defaultJitMethodImage : defaultMethodImage;
+			}
+			return nonoptimizedMethodImage;
+		}
+		return null;
+	}
+
+	@Override
+	public String getText(Object key) {
+		if (key instanceof MethodWithFrameType) {
+			MethodWithFrameType mwft = (MethodWithFrameType) key;
+			key = mwft.getMethod();
+		}
+		if (key instanceof IMCFrame) {
+			key = ((IMCFrame) key).getMethod();
+		}
+		if (key instanceof IDisplayable) {
+			return ((IDisplayable) key).displayUsing(IDisplayable.EXACT);
+		}
+		return TypeHandling.getValueString(key);
+	};
+
+	@Override
+	public String getToolTipText(Object key) {
+		if (key == null) {
+			return "null";
+		}
+		if (key instanceof IDisplayable) {
+			return ((IDisplayable) key).displayUsing(IDisplayable.VERBOSE);
+		}
+		return key.toString();
```

Comment:
When giving this a try I'm finding that the key object is a `MethodWithFrameType`, and as such it doesn't get formatted into any human readable way.

![2021-06-17-151839_1916x1036_scrot](https://user-images.githubusercontent.com/10425301/122470248-d40da580-cf8b-11eb-8286-a3855edf00d7.png)

Reply:
> Is the design of these icons standardized, and is there a place to find out in jmc what they mean? Should this information be added to the tooltip perhaps? Otherwise I added a couple of comments inline, other than those I think it looks good.

Yeah, it's the same geometric figures as Eclipse is using. ;)

Reply:
Good catch with the tooltip! Thanks!